cmd/query/app/querysvc/adjuster/spaniduniquifier.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2019 The Jaeger Authors.
+// Copyright (c) 2017 Uber Technologies, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+)
+
+const (
+	warningTooManySpans = "cannot assign unique span ID, too many spans in the trace"
+)
+
+// SpanIDUniquifier returns an adjuster that changes span ids for server
+// spans (i.e. spans with tag: span.kind == server) if there is another
+// client span that shares the same span ID. This is needed to deal with
+// Zipkin-style clients that reuse the same span ID for both client and server
+// side of an RPC call. Jaeger UI expects all spans to have unique IDs.
+//
+// This adjuster never returns any errors. Instead it records any issues
+// it encounters in Span.Warnings.
+func SpanIDUniquifier() Func {
+	return Func(func(traces ptrace.Traces) error {
+		adjuster := spanIDDeduper{
+			spansByID: make(map[pcommon.SpanID][]ptrace.Span),
+			maxUsedID: pcommon.NewSpanIDEmpty(),
+		}
+		return adjuster.adjust(traces)
+	})
+}
+
+type spanIDDeduper struct {
+	spansByID map[pcommon.SpanID][]ptrace.Span
```

Comment:
```suggestion
	spansByID map[pcommon.SpanID][]*ptrace.Span
```
You don't want to be copying spans by value here.

Reply:
@yurishkuro is this needed because the Span type is just a wrapper around pointers?
```go
type Span struct {
	orig  *otlptrace.Span
	state *internal.State
}
```

Reply:
+1